A company faces a significant challenge related to frequent plan cancellations. To address this issue, I conducted a comprehensive analysis project, employing both theoretical and experimental approaches. It's important to note that this project is not based on a real-world case but rather serves as an experimental activity aimed at practicing and refining analytical skills.

Throughout this initiative, I identified cancellation trends, analyzed underlying factors, and proposed strategic solutions to effectively mitigate and resolve the problem.

During the analysis process, we outlined recurring cancellation patterns, allowing for a deeper understanding of the reasons behind this trend. Through a proactive approach, we applied detailed research methods and data analysis to extract valuable insights into customer behavior and the inflection points that lead to plan cancellations.

Based on the results, we developed and implemented customized strategies, targeting specific areas identified as prone to high cancellation rates. These solutions were designed not only to contain cancellation rates but also to enhance customer satisfaction, thereby strengthening brand loyalty.

In summary, the analysis project played a crucial role in identifying, understanding, and theoretically and experimentally resolving the plan cancellation problem, contributing to the stability and sustainable growth of the company.

By Antonio David Santana

E-mail: davidantoniodavi2@gmail.com

LinkedIn: linkedin.com/in/antonlo-davld

GitHub: github.com/AntDavid

In [8]:
import plotly.express as px
import pandas as pd

#Import the data base

data_base = pd.read_csv("/content/drive/MyDrive/Colab Notebooks/data_analytics/cancelamentos.csv")
display(data_base)


CustomerID  idade    sexo  tempo_como_cliente  frequencia_uso  \
0              2.0   30.0  Female                39.0            14.0   
1              3.0   65.0  Female                49.0             1.0   
2              4.0   55.0  Female                14.0             4.0   
3              5.0   58.0    Male                38.0            21.0   
4              6.0   23.0    Male                32.0            20.0   
...            ...    ...     ...                 ...             ...   
881661    449995.0   42.0    Male                54.0            15.0   
881662    449996.0   25.0  Female                 8.0            13.0   
881663    449997.0   26.0    Male                35.0            27.0   
881664    449998.0   28.0    Male                55.0            14.0   
881665    449999.0   31.0    Male                48.0            20.0   

        ligacoes_callcenter  dias_atraso assinatura duracao_contrato  \
0                       5.0         18.0   Standard           Annual   
1                      10.0          8.0      Basic          Monthly   
2                       6.0         18.0      Basic        Quarterly   
3                       7.0          7.0   Standard          Monthly   
4                       5.0          8.0      Basic          Monthly   
...                     ...          ...        ...              ...   
881661                  1.0          3.0    Premium           Annual   
881662                  1.0         20.0    Premium           Annual   
881663                  1.0          5.0   Standard        Quarterly   
881664                  2.0          0.0   Standard        Quarterly   
881665                  1.0         14.0    Premium        Quarterly   

        total_gasto  meses_ultima_interacao  cancelou  
0            932.00                    17.0       1.0  
1            557.00                     6.0       1.0  
2            185.00                     3.0       1.0  
3            396.00                    29.0       1.0  
4            617.00                    20.0       1.0  
...             ...                     ...       ...  
881661       716.38                     8.0       0.0  
881662       745.38                     2.0       0.0  
881663       977.31                     9.0       0.0  
881664       602.55                     2.0       0.0  
881665       567.77                    21.0       0.0  

[881666 rows x 12 columns]

In [2]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
#Visually inspect the data base
    #Analyze the data base and remove columns that do not contribute to the analysis.

data_base = data_base.drop(columns="CustomerID")

#Correct errors in the data base and visualize problematic cells.

display(data_base.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 881666 entries, 0 to 881665
Data columns (total 11 columns):
 #   Column                  Non-Null Count   Dtype  
---  ------                  --------------   -----  
 0   idade                   881664 non-null  float64
 1   sexo                    881664 non-null  object 
 2   tempo_como_cliente      881663 non-null  float64
 3   frequencia_uso          881663 non-null  float64
 4   ligacoes_callcenter     881664 non-null  float64
 5   dias_atraso             881664 non-null  float64
 6   assinatura              881661 non-null  object 
 7   duracao_contrato        881663 non-null  object 
 8   total_gasto             881664 non-null  float64
 9   meses_ultima_interacao  881664 non-null  float64
 10  cancelou                881664 non-null  float64
dtypes: float64(8), object(3)
memory usage: 74.0+ MB


None

In [ ]:
#Remove null cells

data_base = data_base.dropna()
display(data_base.info())

<class 'pandas.core.frame.DataFrame'>
Int64Index: 881659 entries, 0 to 881665
Data columns (total 11 columns):
 #   Column                  Non-Null Count   Dtype  
---  ------                  --------------   -----  
 0   idade                   881659 non-null  float64
 1   sexo                    881659 non-null  object 
 2   tempo_como_cliente      881659 non-null  float64
 3   frequencia_uso          881659 non-null  float64
 4   ligacoes_callcenter     881659 non-null  float64
 5   dias_atraso             881659 non-null  float64
 6   assinatura              881659 non-null  object 
 7   duracao_contrato        881659 non-null  object 
 8   total_gasto             881659 non-null  float64
 9   meses_ultima_interacao  881659 non-null  float64
 10  cancelou                881659 non-null  float64
dtypes: float64(8), object(3)
memory usage: 80.7+ MB


None

In [ ]:
#Initial analysis of the problem

#How many people canceled the plan (numerical value)?

display(data_base["cancelou"].value_counts ())

#How many people canceled the plan (percentage value)?

display(data_base["cancelou"].value_counts(normalize=True).map("{:.2%}".format))

1.0    499993
0.0    381666
Name: cancelou, dtype: int64

1.0    56.71%
0.0    43.29%
Name: cancelou, dtype: object

In [ ]:
#Analysis of the causes of the problem (how the columns in the data base impact the issue)

#Format colors of the graphs

color_map = {1: 'yellow', 0: 'blue'}

#Create Graphs

for coluna in data_base.columns:
    grafico = px.histogram(data_base, x=coluna, color="cancelou",  color_discrete_map=color_map)
    grafico.show()

Output hidden; open in https://colab.research.google.com to view.

In [ ]:
#All customers with monthly contracts are canceling
    #Offer discounts on other plans

data_base = data_base[data_base["duracao_contrato"]!="Monthly"]

#Returns the new cancellation value applying the condition

display(data_base["cancelou"].value_counts())

#Returns the new cancellation percentage applying the condition

display(data_base["cancelou"].value_counts(normalize=True).map("{:.2%}".format))

0.0    381666
1.0    325788
Name: cancelou, dtype: int64

0.0    53.95%
1.0    46.05%
Name: cancelou, dtype: object

In [ ]:
#Customers who call the call center more than 4 times cancel
    #Create a process to resolve customer issues in a maximum of 4 calls (avoid callbacks, preferably)

data_base = data_base[data_base["ligacoes_callcenter"]<=4]

#Returns the new cancellation value applying the condition

display(data_base["cancelou"].value_counts())

#Returns the new cancellation percentage applying the condition

display(data_base["cancelou"].value_counts(normalize=True).map("{:.2%}".format))

0.0    379032
1.0    136016
Name: cancelou, dtype: int64

0.0    73.59%
1.0    26.41%
Name: cancelou, dtype: object

In [ ]:
#Customers who delay payment by more than 20 days cancel
    #Resolve delays within 10 days

data_base = data_base[data_base["dias_atraso"]<=20]

#Returns the new cancellation value applying the condition

display(data_base["cancelou"].value_counts())

#Returns the new cancellation percentage applying the condition

display(data_base["cancelou"].value_counts(normalize=True).map("{:.2%}".format))

0.0    379032
1.0     85447
Name: cancelou, dtype: int64

0.0    81.60%
1.0    18.40%
Name: cancelou, dtype: object

In [ ]:
#All customers over 50 years old cancel
    #Apply resources to retain such customers and/or define younger people as target audience

data_base = data_base[data_base["idade"]<=50]

#Returns the new cancellation value applying the condition

display(data_base["cancelou"].value_counts())

#Returns the new cancellation percentage applying the condition

display(data_base["cancelou"].value_counts(normalize=True).map("{:.2%}".format))

0.0    379032
1.0     52121
Name: cancelou, dtype: int64

0.0    87.91%
1.0    12.09%
Name: cancelou, dtype: object

In [ ]:
#All customers who spent less than $500 canceled
    #Increase the monthly plan prices to encourage customers to choose quarterly and annual plans,
    #  thereby increasing their spending

data_base = data_base[data_base["total_gasto"]>=500]

#Returns the new cancellation value applying the condition

display(data_base["cancelou"].value_counts())

#Returns the new cancellation percentage applying the condition

display(data_base["cancelou"].value_counts(normalize=True).map("{:.2%}".format))

0.0    379032
1.0     19134
Name: cancelou, dtype: int64

0.0    95.19%
1.0     4.81%
Name: cancelou, dtype: object

In [ ]:
#Display the results of the implementation of the solutions

#Format colors of the graphs

color_map = {1: 'red', 0: 'gray'}

#Create Graphs

for coluna in data_base.columns:
    grafico = px.histogram(data_base, x=coluna, color="cancelou",  color_discrete_map=color_map)
    grafico.show()

Output hidden; open in https://colab.research.google.com to view.

Before implementation:

Canceled:

Yes: 56.71%

No: 43.29%

After implementation:

Canceled:

Yes: 4.81%

No: 95.19%

In this theoretical case, we observe that the implementation helped solve the problem, resulting in better customer retention. As a former physics student, there's an interesting joke we often used that fits well in this situation: "When a mechanics experiment fails, it's because the physicist habitually neglected friction" Therefore, in this situation, I understand that there may be challenges hindering the implementation of all solutions, depending on the financial situation of the company, and it requires time and coordination across departments. However, I believe the solutions are applicable, and in a real-world scenario, they can be implemented based on the company's needs and financial conditions.